packages/amplify-codegen-e2e-tests/src/__tests__/add-codegen.test.ts

```diff
@@ -0,0 +1,116 @@
+        // graphql configuration should not be added
+        expect(existsSync(path.join(projectRoot, graphqlConfigFile))).toBe(false);
+    });
+
+    it(`Adding codegen twice gives appropriate message in JS project`, async () => {
+        const config = DEFAULT_JS_CONFIG;
+        const pathToConfigFile = path.join(projectRoot, graphqlConfigFile);
+
+        // init project and add API category
+        await initProjectWithProfile(projectRoot, { ...config });
+        await addApiWithSchema(projectRoot, schema);
+
+        // generate pre-existing user file
+        const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);
+        // given that the pre-existing file exists
+        expect(existsSync(userSourceCodePath)).toBe(true);
+
+        // adding codegen succeeds and generates a config file
+        await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();
+        expect(existsSync(pathToConfigFile)).toBe(true);
+        
+        // adding codegen again gives appropriate message
+        const settings = { isCodegenAdded: true, ...config }; 
+        await expect(addCodegen(projectRoot, settings)).resolves.not.toThrow();
+
+        // pre-existing file should still exist
+        expect(existsSync(userSourceCodePath)).toBe(true);
+        // previously generated files should still exist
+        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
+        // check if the added codegen configuration is valid
+        expect(readFileSync(pathToConfigFile).toString()).toMatchSnapshot();
+    });
+
+    frontendConfigs.forEach(config => {
+      it(`Adding codegen works as expected in ${config.frontendType} project`, async () => {
+        // init project and add API category
+        await initProjectWithProfile(projectRoot, { ...config });
+        await addApiWithSchema(projectRoot, schema);
+
+        // generate pre-existing user file
+        const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);
+        // given that the pre-existing file exists and codegen output directory is empty
+        expect(existsSync(userSourceCodePath)).toBe(true);
+        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(false);
+
+        // add codegen succeeds
+        await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();
+
+        // pre-existing file should still exist
+        expect(existsSync(userSourceCodePath)).toBe(true);
+        // GraphQL statements are generated
+        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
+        // graphql configuration should be added
+        expect(existsSync(path.join(projectRoot, graphqlConfigFile))).toBe(true);
+        // check if the added codegen configuration is valid
+        expect(readFileSync(path.join(projectRoot, graphqlConfigFile)).toString()).toMatchSnapshot();
+      });
+    });
```

Comment:
These tests are the same as the tests in `graphql-codegen` except for the snapshots. https://github.com/aws-amplify/amplify-codegen/blob/f0e6857d84d2e7fe91969881a9aadaeac5c9e533/packages/amplify-codegen-e2e-tests/src/__tests__/graphql-codegen.test.ts#L41-L51

Reply:
Testing various user flows in a single test suite might not be optimal. We can reduce the time taken for running the e2e tests by running each frontend being tested in parallel and also splitting the test suites into a separate job in CircleCI. I will include the circleci changes in a separate PR.